Elasticsearch configurable and with management scripts

boyska opened the pull request:
The first commit is about #61 : making the index_name configurable, so that you can run multiple webant webservers, with different dbs, on a single elasticsearch instance. The option is `ES_INDEXNAME`

Since it changes some default (before, the hardcoded index was `book`; now there's a default one, which is `libreant`), I thought it was a good moment to create some nice utility to export/import/search things from the db.

For example, this is a "migration" script from `book` to `libreant`
```sh
WEBANT_ES_INDEXNAME=book libreant-manage db_export | libreant-manage db_import -
```

You can do searches with `db_search`, export single items with `db_item_export` and import them with `db_import`.

Want to do a complete dump of your db? `libreant-manage db_export > dbdump.json` is what you need.

Implementation is based on Flask-Script, which is a very common extension to make the creation of commands very easy. You can see that `manage.py` is actually quite simple, and adding new commands should be very easy.

Also, Flask-Script comes bundled with a nice "shell" command, which will just open a shell. If you install ipython (`pip install ipython`), the shell will be more interactive, with tab_completion and other niceties: really helpful when coding or debugging

boyska commented:
Other coolness that is now possible:
```sh
curl -s 'http://boyska.s.pt-labs.net/libreant/contrib/colibri.json' | WEBANT_ES_INDEXNAME=colibri webant-manage db_import -
curl -s  'http://boyska.s.pt-labs.net/libreant/contrib/rabbia/rabbia.json' | WEBANT_ES_INDEXNAME=rabbia webant-manage db_import -
curl -s   'http://boyska.s.pt-labs.net/libreant/contrib/csv2json/info_forte.json' | WEBANT_ES_INDEXNAME=forte webant-manage db_import -

WEBANT_PORT=5002 WEBANT_ES_INDEXNAME=colibri webant
WEBANT_PORT=5001 WEBANT_ES_INDEXNAME=rabbia webant
 WEBANT_AGHERANT_DESCRIPTIONS='["http://localhost:5001/description.xml", "http://localhost:5002/description.xml"]' agherant
```

and you'll quickly setup some nodes, with realistic DBs, plus one node that is just an aggregator.

ael-code commented on webant/webant.py:
```diff
@@ -71,6 +72,8 @@ def get_db():
             app._db = db
         return app._db
 
+    app.get_db = _get_db
```
find & replace?

boyska replied:
find&replace of what?
if you're wondering why add get_db as part of app, it's because otherwise `manage.py` (and many similar things) could never work.

Previously `get_db` was a very strange fact: a function defined in a not-global scope, but not part of the app itself.

ael-code commented on setup.py:
```diff
@@ -106,6 +107,7 @@ def read(fname):
                 'compile_translations': compile_translations },
       entry_points={'console_scripts': [
           'webant=webant.webant:main',
+          'webant-manage=webant.manage:main',
```
why did you choose this name?
the application name is libreant. The web interface module name is webant.
It's very strange that the command to menage libreantdb it's called webant-manage.

boyska replied:
I had the same doubt, so here's why:
* it heavily depends on webant
* we could want to include in it some web-specific subcommands

Despite this reasons, I am not very sure about the name.

ael-code commented on libreantdb/api.py:
```diff
@@ -40,9 +40,9 @@ class DB(object):
     '''
     # Setup {{{2
-    def __init__(self, es):
+    def __init__(self, es, index_name='book'):
```
default index name should be ```libreant``` in order to be consistent with the default choosen in webant module and because we do not store only book!

ael-code commented on libreantdb/api.py:
```diff
@@ -107,6 +107,10 @@ def setup_db(self):
     # Queries {{{2
+    def __len__(self):
+        stats = self.es.indices.stats()
+        return stats['indices']['book']['total']['docs']['count']
```
hardcoded index name?

boyska replied:
this is a bug. will fix soon, thanks

ael-code commented on webant/manage.py:
```diff
@@ -0,0 +1,70 @@
+
+
+@manager.command
+def db_import(filename, language='it'):
```
default language ```it```?

boyska replied:
`@manager.command` will automatically make an option out of keyword arguments. Try it! It will create `-l`, or `--language`, so you can specify the language manually.

ael-code commented:
I like the new command possibilities. It is very cool.
But I did not understand why a command to manage libreant database is located under webant module and need to launch webant to perform all the tasks.
This is not a modular way to handle code.
We could move this command into libreantdb and use some other library to parse command line argument?

boyska commented:
Short answer: no, we can't.

I mean, we could. But the cool thing about this is that it accepts the same configuration (and has therefore a much similar behaviour) to webant. Furthermore, it can deal with everything managed by weband, so it's not libreantdb-specific. Some examples:
* we could (and probably should!) implement a `rich_item_export` subcommand that will export not only the elasticsearch item, but even the files attached to it. This cannot fit into libreantdb, because it needs to access fsdb.
* when we'll have proper cache support, we might want to have some command to do common cache operation: flush, show cache size, list cached things, etc. This will be particularly web-specific


Another small clarification: this never "launches" webant. It just uses the same application code that is also exposed by WSGI.

I think that we are discussing is actually something about model-view-controller separation. To implement what you say we'd have to move all the application logic into an explicit controller, which the web part and the commandline part could then use. But this is clearly *not* the approach of flask!
Apart from the "routes" there's nothing view-specific in the application.

So, here's some proposals:
1. we could just keep it how it is. It works, and is not SO bad
2. we could subclass the Flask type. [Subclassing is a pattern](http://flask.pocoo.org/docs/0.10/becomingbig/#subclass) for getting flask more structured. We could move everything application-specific into a new class called LibreantCoreApp (get_db, fsdb, cache, user handling, etc) and keep views, Bootstrap and more web-related things just inside create_app, or in a subclass of LibreantCoreApp called LibreantViewApp

ael-code commented:
A lot of operations could be served as REST api. (#52)
I did not exactly understand the second choice you provided. Maybe we have to talk about it and take a good decision.

boyska commented:
A REST API is not a commandline interface, anyway ;) How are you going to programmatically use it? With a script involving curl, probably. This just moves the problem a bit further, but creates new problems.

HTTP is sometimes nice, but it just does not fit the case of:
* long-running jobs
* a quick superadmin way of doing things. If you have to login in a browser, running scripts become way more difficult
* it could be preferred to not expose some functions over http at all, not even to site admin. Some things are just for sysadmins!
* simplicity: why should I need to run a webserver just to do a db export?

boyska commented:
About the subclassing proposals: if you want, I can give a quick example of what I mean, in a different branch.

ael-code commented:
> About the subclassing proposals: if you want, I can give a quick example of what I mean, in a different branch.

Oke

boyska commented:
Okay, you can look it in https://github.com/boyska/libreant/commits/subclass

The last commit is a bit pointless, since routes have not been moved to LibreantViewApp just because I was lazy. The first commit (490c8077) is the important one that would solve your concerns. Note, also, that some application logic could/should be moved to LibreantCoreApp.
This is the case, for example, of the upload function: the code dealing with fsdb and libreantdb could be put inside LibreantCoreApp, so that both the web part and the manage scripts can use it nicely.
The last one (853455c100) just serves to illustrate what subclassing can do for us.

ael-code commented:
oke it's more simple then i thouth.

boyska commented:
so what's your opinion?

boyska commented:
This would close #74

boyska commented:
I addressed all your considerations:
* the script is now called `libreant-manage`, which makes more sense.
* the experiment on subclassing flask has now been integrated, so there is more separation between application logic and view stuff
* the bug about `'book'` hardcoded has been fixed, and a unit test about it has been added
* you said that having `index_name='book'` was a bad idea, and proposed `libreant` as a better default. I decided to give no default at all. Libreant still has a default value, but the libreantdb does not. I think it's more meaningful this way: have you ever seen a class for opening  a mysql db with default uri and db? ;)

So I believe that this is ready and actually makes things even tidier than they were. `LibreantCoreApp` is a good place to put more application logic, making the view more and more "stupid".

frenko commented:
I do not know if what I say is related to this problem but if I create two indices (Colibrì and Rabbia) and then do a search on agherant with ``search?q=ciao`` I always get 0 results. Even if I try all the results with ``search?q=*:* `` I always get 0. I have something wrong?

boyska commented:
@frenko Do you have both nodes running, right? (just open three terminals, 2 for the libreant nodes and one for agherant)

frenko commented:
@boyska yes, are running both

boyska commented:
You should see something like this https://i.imgur.com/RmUfhrq.png
I mean: browser request arriving to agherant, then requests to every other node.

By the way, is it `/agherant/search?q=*:*`, right?

frenko commented:
>I mean: browser request arriving to agherant, then requests to every other node.

I don't know because I have run webant with "&" , I will try!

>By the way, is it /agherant/search?q=*:*, right?

Yes!

boyska commented:
On Fri, Jan 23, 2015 at 07:02:15AM -0800, Frenko wrote:
>>I mean: browser request arriving to agherant, then requests to every other node.
>
>I don't know because I have run webant with "&" , I will try!

this doesn't change a thing, I believe

frenko commented:
If I send a request to agherant the two instances of webant not receive any requests!!
![screenshot from 2015-01-23 16 52 46](https://cloud.githubusercontent.com/assets/6986802/5877500/466a0942-a320-11e4-9039-2486841a0a22.png)

boyska commented:
Is 10.10.10.3 your IP? if so, why are you receiving requests from 10.10.10.1?

What if you manually try to visit the description url you provided inside the descriptions? will they work?

frenko commented:
the ip of virtual machine is 10.10.10.3, ip of Host pc is 10.10.10.1. I use host-only network to connect to guest machine. If I get request from host my IP is 10.10.10.1.
From description.xml:
``` xml
<OpenSearchDescription>
<ShortName>Web Search</ShortName>
<Description>Use Example.com to search the Web.</Description>
<Tags>example web</Tags>
<Contact>admin@example.com</Contact>
<Url type="application/rss+xml" template="http://10.10.10.3:5002/search?format=application%2Frss%2Bxml&q={searchTerms}"/>
<Url type="text/html" template="http://10.10.10.3:5002/search?format=text%2Fhtml&q={searchTerms}"/>
<OutputEncoding>UTF-8</OutputEncoding>
<InputEncoding>UTF-8</InputEncoding>
</OpenSearchDescription>
```

boyska commented:
what happens if you use localhost instead of 10.10.10.3? (I know that 10.10.10.3 is the correct thing, but just to be sure).

Moreover: what happens if, from **inside** the vm you run `curl http://10.10.10.3:5002/description.xml` ?

frenko commented:
>what happens if you use localhost instead of 10.10.10.3? (I know that 10.10.10.3 is the correct thing, but just to be sure).

the same thing!

>Moreover: what happens if, from inside the vm you run ``curl http://10.10.10.3:5002/description.xml`` ?

I obtain the same output!

boyska commented:
And can you visit http://10.10.10.3:5001/ from your browser? does it work?

it's really strange!

frenko commented:
instances of webant for colibrì and rabbia work perfectly!
I don't know what's the matter

boyska commented:
ok, I believe that this is a separate issue from this pull request :)

Please open a bug about it; I will try to add more debug information to agherant.

boyska commented:
This still hasn't been merged. Is there any blocking issue? can we merge? @frenko I don't think that this pull request created the problem you are experiencing. Could you test the same thing on insomnia-lab/master ? If the bug is also present there, then it's a bug, and is not related to this pull request. If it's related to this pull request, it's a blocker.

ael-code commented:
I would like to personal review this pull request.
Can you wait for me? :)

boyska commented:
sure I can wait :)

ael-code commented:
> I do not know if what I say is related to this problem but if I create two indices (Colibrì and Rabbia) and then do a search on agherant with search?q=ciao I always get 0 results. Even if I try all the results with search?q=*:* I always get 0. I have something wrong?

The problem it is about a bug in agherant_standalone module initialization.
AppConfig use the flask app name as prefix of environment configuration names.
We initialize Flask app on agherant_standalone using classic ```Flask(__name__)``` call, but in this case ```__name__``` is equal to ```WEBANT.AGHERANT_STANDALONE```.
The result is that all the environment configurations prefixed with ```WEBANT``` are ignored.

boyska commented:
wow @ael-code that's a very good point! I think that just hardcoding `"webant"` is the correct thing to do in this case.

anyway, this bug is not introduced by this pull request, so this should be fixed separately (#81)

ael-code commented:
I started to test these changes, but I need some time to test them all :)
Anyway I like the new modularized template.

ael-code commented on webant/webant.py:
```diff
@@ -18,7 +18,52 @@
+            'DEBUG': True,
+            'PRESET_PATHS': [],  # defaultPreset should be loaded as default?
+            'FSDB_PATH': "",
+            'AGHERANT_DESCRIPTIONS': [],
```
I don't like that ```AGHERANT_DESCRIPTIONS```, ```PRESET_PATHS``` and other child-class related configuration are initialized in LibreantCoreApp.
We can add a dict parameter to the ```__init__``` function in order to pass defaults from sub-classes.

boyska replied:
seems correct. will try to write this in the afternoon

boyska replied:
ok, corrected. I'm not sure that this is what you mean, but I am sure that this is what I mean :)

Just in case you were thiinking that no default should be in LibreantCoreApp, so the logic can follow one of these options:
1. each parameter HAS to be specified by the initializer. Which is uncommon in python things, whereas sane defaults are much more easy to deal with
2. set default values inside the class, then override them after having it instantiated. This will create "db connections" point to wrong uri (fsdb, elasticsearch, whatever; all would be wrong)

Doing the change you asked, I also moved the option about bootstrap to LibreantViewApp, which is more correct, because LibreantCoreApp does not care about it.

ael-code replied:
> ok, corrected. I'm not sure that this is what you mean, but I am sure that this is what I mean :)

This is not what I mean.
The problem is that you have reorganized the code to make it more modularized but you did not the same with configuration stuff.

The big question is:
Why ```LibreantCoreApp``` handle default setting values for something that it will never use?
For example ```AGHERANT_DESCRIPTIONS``` will be used only by the sub-class ```LibreantViewApp``` but the default value is initialized in ```LibreantCoreApp```.

The pattern that I suggest is something like this:

```LibreantCoreApp``` make a first call to ```config.update()``` to initialize default values that are going to be used in this class (```DEBUG```,```FSDB_PATH```,```SECRET_key``` etc...).
Then it make another call to ```config.update()``` using as paramter the one received in the constructor in order to set default values for stuff that are going to be used in the child-classes.

Using this pattern ```LibreantCoreApp``` will start like this:
```python
def __init__(self, import_name, defConf={}):
   super(LibreantCoreApp, self).__init__(import_name)
   localDefConfig = { 'DEBUG': True, 'FSDB_PATH': "", 'SECRET_KEY': 'asd', 'ES_INDEXNAME': 'libreant' }
   self.config.update(localDefConfig)
   self.config.update(defConf)
...
```
and  ```LibreantViewApp``` will start like this:

```python
def __init__(self, import_name, defConf={}):
   localDefConf = {'PRESET_PATHS': [], 'BOOTSTRAP_SERVE_LOCAL': True, 'AGHERANT_DESCRIPTION':[] }
   super(LibreantViewApp, self).__init__(import_name, localDefConf.update(defConf))
   ...
```

ael-code commented:
> ok, corrected. I'm not sure that this is what you mean, but I am sure that this is what I mean :)

This is not what I mean.
The problem is that you have reorganized the code to make it more modularized but you did not the same with configuration stuff.

The big question is:
Why ```LibreantCoreApp``` handle default setting values for something that it will never use?
For example ```AGHERANT_DESCRIPTIONS``` will be used only by the sub-class ```LibreantViewApp``` but the default value is initialized in ```LibreantCoreApp```.

The pattern that I suggest is something like this:

```LibreantCoreApp``` make a first call to ```config.update()``` to initialize default values that are going to be used in this class (```DEBUG```,```FSDB_PATH```,```SECRET_key``` etc...).
Then it make another call to ```config.update()``` using as paramter the one received in the constructor in order to set default values for stuff that are going to be used in the child-classes.

Using this pattern ```LibreantCoreApp``` will start like this:
```python
def __init__(self, import_name, defConf={}):
   super(LibreantCoreApp, self).__init__(import_name)
   localDefConfig = { 'DEBUG': True, 'FSDB_PATH': "", 'SECRET_KEY': 'asd', 'ES_INDEXNAME': 'libreant' }
   self.config.update(localDefConfig)
   self.config.update(defConf)
...
```
and  ```LibreantViewApp``` will start like this:

```python
def __init__(self, import_name, defConf={}):
   localDefConf = {'PRESET_PATHS': [], 'BOOTSTRAP_SERVE_LOCAL': True, 'AGHERANT_DESCRIPTION':[] }
   super(LibreantViewApp, self).__init__(import_name, localDefConf.update(defConf))
   ...
```

boyska commented:
I don't believe that agherant only is about "views". Same way I don't believe that preset_path is only about views.

My idea was to separate web-only thing, from things that make sense to be called even without passing from a web application (that is, command line scripts!).
Well, agherant could be useful to be used from command line-scripts. Example
`WEBANT_AGHERANT_DESCRIPTIONS=blabalbal libreant-manage agherant-check-nodes`

Same is true for presets: `libreant skeleton-list` seems a useful thing to me.

But it's not just "because it's useful", it's also (I believe) more correct. For example, we may need some helper functions for agherant, or for presets. Those functions should live in CoreApp, not in View.

Moving those just to "view" will, I believe, make libreant core logic less defined, and much delegated to a view part. Moving as much as possible to core will make the "logic" bigger, and the view thiner. Which could make it easy to change views, create new views, etc. all without touching the semantics of the program.

ael-code commented:
I think this is an important change so I will spend some more words.

> My idea was to separate web-only thing, from things that make sense to be called even without passing from a web application (that is, command line scripts!).
> Well, agherant could be useful to be used from command line-scripts. Example
> WEBANT_AGHERANT_DESCRIPTIONS=blabalbal libreant-manage agherant-check-nodes
> 
> Same is true for presets: libreant skeleton-list seems a useful thing to me.

I am trying to understand...

You want to use agherant from within ```libreant-manage```, but were are you going to initialize it? As of now agherant is initialize only in ```LibreantViewApp```.

boyska commented:
>> My idea was to separate web-only thing, from things that make sense to be called even without passing from a web application (that is, command line scripts!).
>> Well, agherant could be useful to be used from command line-scripts. Example
>> WEBANT_AGHERANT_DESCRIPTIONS=blabalbal libreant-manage agherant-check-nodes
>>
>> Same is true for presets: libreant skeleton-list seems a useful thing to me.
>
>I am trying to understand...
>
>You want to use agherant from within ```libreant-manage```, but were
>are you going to initialize it? As of now agherant is initialize only
>in ```LibreantViewApp```.

ATM we never initialize agherant, because there is simply no
initializaion needed (oh, maybe you can say that we initialize right in
agherant.py)

ATM agherant is very simple, and does not offer any API. But it will!
And if it will need some initialization, I think that LibreantCoreApp
will be the place to do it.

ael-code commented:
> ATM we never initialize agherant, because there is simply no initializaion needed (oh, maybe you can say that we initialize right in agherant.py)

with **initialize** I mean [ this two line ](https://github.com/insomnia-lab/libreant/blob/master/webant/webant.py#L46-L47).

Anyway **we are talking about different things**:
You are talking about the right place to put default config and I am talking about putting initialization code and default config in the same class.

If we decide to initialize presetManager in ```LibreantCoreApp``` we should put also default config for presetManager in ```LibreantCoreApp```.

But at the moment we initialize presetManager inside ```LibreantViewApp``` so we should manage default config for presetManager inside ```LibreantViewApp```.

Is going to be very diffcult to explain myself.

boyska commented:
That is not initialization, anyway. This is adding the Blueprint to the application, which is a view-specific thing.

About presetmanager, you are completely right: initialization should be moved to LibreantCoreApp, I believe. Yes, at this point it's not completely coherent. I will be glad both to merge your code about this, or to write it myself. But I'd like to know if you think that those issues are blockers for this pull request, or if instead it can be merged and we can reorganize it later.

ael-code commented:
If we are going to separate configuration handling from flask-app, I think that we could make a big change:

make a class ```LibreantCore``` that is indipendent from flask. 
Inside this class we could manage configurations and initialize related module like  PresetManager, fsdb, Esdb etc...
This class will be used from cli through ```libreant-manage``` and on the other hand accessible from flask-app.

I like this code organization because:
 - we won't need to start flask to use ```libreant-manage```
 - the code is very modular. We could even change web framework without touching ```LibreantCore``` stuff.


Yes... we need to change library to handle ```libreant-manage```. Don't be lazy!!!

boyska commented:
mh, not truly convinced (while tempted). The change in librant-manage will be trivial, so I am not concerned by this. There are some other points that I don't really like:
* if we do like this, every module (ok, agherant; we have no other, by now) should be splitted in a non-flask part and one flask-specific part. Isn't this too much overhead for a code that is not so complicated?
* There is no such thing as "starting flask". Flask is not a server, in case you are wondering :) I don't think that being completely agnostic in every part is the more correct thing to do. It can lead to overcomplexity, with no clear advantage. Let's, for example, talk of agherant. At now, it is very simple. It will soon need some concurrency method (see #85), `gevent` probably being the choice. Should we be "modular", providing a layer of abstraction for "a pluggable concurrency method"? I don't think that we should go _that_ modular. Same applies for `flask`: it's a microframework, let's use it.
* Decoupling everything from flask will make something difficult. For example, we'll soon need to connect to some sql database. We still haven't reached a decision, but a good candidate is to use sqlalchemy, and the [flask-sqlalchemy](http://flask.pocoo.org/docs/0.10/patterns/sqlalchemy/) could be a good solution. It is a flask-extension, so could not fit in the core application, in theory. So we would be forced to not use it, making integration of other authentication-related parts even more difficult. If we were really needing it, I would agree, it's not impossible. But is it really so important?

ael-code commented:
I really don't like the actual all-in-flask organization. 
For me making open source code means also that we must provide modular, reusable and easy-to-change code otherwise became open-source-but-untouchable.

All your point says "it will be more complex" and you are right. But I tihnk that making the hard part now make things easy for the future and for others.

Anyway the code works fine in both way so we can continue with the fully-flasked way :(

boyska commented:
Your idea is much correct, but I really don't know how to implement it without getting crazy (making code complicated is another way to make open-source-but-untouchable, btw ') ). 
If you have some idea on a sane way to make code more modular, let's try to!

boyska commented:
any news on this?

ael-code commented:
I'm waiting for your update.
Points to be solved:
 - configuration pattern does not work as expected. ```conf``` parameter coming from a child of ```LibreantViewApp`` is ignored. ( see my previous comment )

 - preset and agherant configurations need to be moved to ```LibreantViewApp```

ael-code commented:
There is another problem. ```initLogger``` call is placed under flask app creation so the log raised during creation are suppressed. 
Is it possible to see log in libreant-manage at the moment? I did not tested it.

boyska commented:
Ok thanks for the summary, I've fixed these things. I followed the principle "even if maybe an option should belong to the parent, if the parent does nothing with it, then it belongs to the child"
* `AGHERANT_DESCRIPTIONS` is now a configuration option of LibreantViewApp. At some point we may find that it is better to move it to the parent (that is, when it will have an API). At this point, there's no need for.
* `PRESET_PATHS` still belong to the parent, who now has the initialization code, too. This could be useful for the CLI version of the skeleton-manager, for example (we discussed a bit about it, but not on github so I cannot link the thread)
* the configuration "precedence" problem you pointed out has been fixed
* `libreant-manage` now has logs.

At the moment, initLoggers decide how much to be verbose based on `app.config['DEBUG']`, so it cannot be run before app has been created. To make it more clear, I rewrote it to take app as an argument.

ael-code commented:
> At the moment, initLoggers decide how much to be verbose based on app.config['DEBUG'], so it cannot be run before app has been created. To make it more clear, I rewrote it to take app as an argument.

If you do not move initLogger before flask app creation, we will lose all boot-logs ( e.g. presetManager validation ).

boyska commented:
On Mon, Feb 09, 2015 at 05:23:23AM -0800, ael wrote:
>If you do not move initLogger before flask app creation, we will lose all boot-logs ( e.g. presetManager validation ).

I know, but how can I fix it? It depends on knowing the debug level,
which is discovered during initialization itself.
This can be solved in two ways, I believe:
* reorganize the config code and handle it ourselves, without delegating
  to AppConfig. See also
  https://github.com/insomnia-lab/libreant/issues/81#issuecomment-73242933
  about this.
* at first set logging at WARNING level, then, if debug is set, set
  logging level to DEBUG.
The two approaches are *not* mutually exclusive, and will just mitigate
a matter of fact: there is alwyas some line of code that cannot use logging.

The bug you report is correct, but is not related to the issues I am
trying to fix (#61 and #74), and is not a regression (it was present
even before this pull request).
You are welcome to open a new issue for that, and I'll also be glad in
fixing it if noone else wants to.

ael-code commented:
> The bug you report is correct, but is not related to the issues I am trying to fix (#61 and #74), and is not a regression (it was present even before this pull request).

Before this pull request, boot-logs were handled fine. Now they are suppressed. Am I wrong?

boyska commented:
On Mon, Feb 09, 2015 at 05:38:06AM -0800, ael wrote:
>> The bug you report [...] is not a regression (it was present even before this pull request).
>
>Before this pull request, boot-logs were handled fine. Now they are suppressed. Am I wrong?

mh, you are partly right. While the chicken-and-egg problem still was
there, very few code was called before initLoggers.
More specifically, fsdb and similar stuff were called after
`initLoggers` so yes, their messages were displayed. I'm not afraid of
fixing it, I am just afraid of turning this issue into
lets-rewrite-everything-about-libreant. If you think this bug is a
blocker, I'll insert the code to change the config handling inside this
pull request, so that we can fix this problem.
Otherwise, I promise I'll do it soon, but after this pull request is
closed.

ael-code commented:
> More specifically, fsdb and similar stuff were called after `initLoggers` so yes, their messages were displayed.

Maybe I'm wrong but it seems to me that all the code inside ```LibreantCoreApp``` and ```LibreantViewApp``` is executed before initLogger call. am I missing something about parallel execution?

boyska commented:
You are right, and there is nothing going parallel. I was talking of the previous behaviour, admitting that there have been a regression.

ael-code commented:
Oke :) English is not my best friend.

boyska commented:
I think that this is what you meant. Since configuration is something that I failed to handle properly in the past, I have done extensive unit-testing for the new `config_utils` module. Hope that this can be the start to more testing inside the `webant` package!

We now have one less dependency, logging starts sooner, and the configuration handling is much more explicit.

boyska commented:
slight change: now, in case of parsing error, the exception is shown, using
```python
log.info(msg, exc_info=1
```

ael-code commented:
Tomorrow I will test all together and I will probably merge this one ;)

ael-code commented:
If debug mode is on, logging message lower then logging.WARNING are suppressed.
I cannot understand where is the problem and I did not find a solution.

Strange fact: if you try to change the default logging level of ```initLogger``` to ```logging.DEBUG``` everything run as expected.

boyska commented:
Mh, can you provide some more details? actually I see debug messages. The only part where I cannot see debug and info messages is in config_utils. But this is because that is executed _before_ the second `initLoggers`.

ael-code commented:
ahaha ok, I screwed up something before. Now they are displayed. Sorry

boyska commented:
ahahah I hate you!

ael-code commented:
No the problem is there...
I checked-out your branch, activated venv, ```python setup.py develop``` and then ```webant```. I do not see fsdb debug message. Do you?

boyska commented:
ok, you fixed it yourself!

ael-code commented:
every time I try to test this pull request a blocking bug comes out. :(

the problem now is that ```libreant-manage``` do not accept config at all.
It is very difficult do not note it, if you launch libreant-manage the following error raises and the cli will crash.

```
ValueError: FSDB_PATH cannot be empty
```

boyska commented:
Bug confirmed. That's because the new config system is not so transparent to tools using it. I'll try to fix it immediately

boyska commented:
OK, that should have been fixed. I also added two subcommands to let you inspect configuration used by libreant-manage (which should be the same used for webant!). I needed those for debugging, and I decided not to remove them, because you could need to debug the same thing! They're on a separate commit so that you can just refuse them if you so prefer.

ael-code commented:
Sorry I notice last update only few minutes ago.
I tested it a little bit and seems to works fine :)
just two more thing:
 - there is no way to understand that using ```db_item_import``` with ```-``` as file name will read input from stdin
 - in both ```db_export``` and ```db_item_export``` debug messages could be a problem (chain and script usage). We can solve it with a verbose/debug flag? If that flag is not trigger log messages are suppressed.

Anyway cli interface of flask-script is very uncomprehensive and not unix-standard-compliant except that I love ```libreant-manage```!

Good work man

boyska commented:
On Fri, Feb 13, 2015 at 03:42:32AM -0800, ael wrote:
> there is no way to understand that using ```db_item_import``` with ```-``` as file name will read input from stdin

well, that's very standard... anyway, I've fixed it explicitly saying
how it works in the help

> - in both ```db_export``` and ```db_item_export``` debug messages could be a problem (chain and script usage). We can solve it with a verbose/debug flag? If that flag is not trigger log messages are suppressed.

they're not a problem, since the default handler goes to stderr. So
piping is not a problem. Check it yourself with
`libreant-manage db_export > /dev/null`

Also, there's no need for a debug flag: the usual WEBANT_DEBUG=false
will do it in a more coherent way (same config method for webant and
libreant-manage)

>Anyway cli interface of flask-script is very uncomprehensive and not
>unix-standard-compliant except that I love ```libreant-manage```!

can't understand what you mean. I think it's the usual way that commands
based on subcommands work. Like git, for example. And all python code
based on argparse (which is very recommended).

I think you mostly refer to the fact that `-h` does not do what you'd
expect, and only `--help` or (crazy) `-?` works. That's a bit weird,
indeed.

ael-code commented:
> they're not a problem, since the default handler goes to stderr. So piping is not a problem. Check it yourself with `libreant-manage db_export > /dev/null` Also, there's no need for a debug flag: the usual WEBANT_DEBUG=false will do it in a more coherent way (same config method for webant and libreant-manage)

ok ok I did not mind about stderr.

> can't understand what you mean. I think it's the usual way that commands based on subcommands work. Like git, for example. And all python code based on argparse (which is very recommended). I think you mostly refer to the fact that `-h` does not do what you'd expect, and only `--help` or (crazy) `-?` works. That's a bit weird, indeed.

I referred to the format of the help message. It is something very unusual to me:

```
usage: libreant-manage [-?]
                       
                       {conf_show,db_item_export,db_import,db_export,conf_list,shell,runserver,db_search}
                       ...

positional arguments:
  {conf_show,db_item_export,db_import,db_export,conf_list,shell,runserver,db_search}
    conf_show           Show the value for a given configuration key
    db_item_export      exports a single item
    db_import           imports items from filename
    db_export           outputs an exportable version of the db
    conf_list           List every key,value pair in configuration
    shell               Runs a Python shell inside Flask application context.
    runserver           Runs the Flask development server i.e. app.run()
    db_search           query, just like you would do in a browser

optional arguments:
  -?, --help            show this help message and exit
```

boyska commented:
I love you.
